src/libraries/Microsoft.PowerFx.Connectors/ConnectorFunction.cs

```diff
@@ -39,58 +41,119 @@ public class ConnectorFunction
+                                                                                                            : new ConnectorParameterWithSuggestions(rp, knownParameters.Skip(ArityMax - 1).ToArray())).ToArray();
 
-            if (httpClient != null)
+            if (HasServiceFunction)
             {
-                var httpInvoker = new HttpFunctionInvoker(httpClient, HttpMethod, OperationPath, ReturnType, ArgumentMapper, cache);
-                invoker = new ScopedHttpFunctionInvoker(DPath.Root.Append(DName.MakeValid(functionName, out _)), Name, functionName, httpInvoker);
+                int index = Math.Min(knownParameters.Length, _serviceFunction.MaxArity - 1);
+                List<RecordValue> suggestions = _serviceFunction.GetConnectorSuggestionsAsync(knownParameters, knownParameters.Length, CancellationToken.None).ConfigureAwait(false).GetAwaiter().GetResult();
+                
+                if (knownParameters.Length >= ArityMax - 1)
+                {
+                    parametersWithSuggestions[index].ParameterNames = suggestions.Select(rv => rv.Fields.First().Name).ToArray();
+                    suggestions = suggestions.Skip(knownParameters.Length - ArityMax + 1).ToList();
+                }
+                
+                parametersWithSuggestions[index].Suggestions = suggestions;
             }
 
-#pragma warning disable SA1117 // parameters should be on same line or all on different lines
-
-            return new ServiceFunction(null, functionNamespace, Name, Name, Description, ReturnType._type, BigInteger.Zero, ArityMin, ArityMax, IsBehavior, false, false, false, 10000, false, new Dictionary<TypedName, List<string>>(),
-                ArgumentMapper.OptionalParamInfo, ArgumentMapper.RequiredParamInfo, new Dictionary<string, Tuple<string, DType>>(StringComparer.Ordinal), "action", ArgumentMapper._parameterTypes)
+            return new ConnectorParameters()
             {
-                _invoker = invoker
+                IsCompleted = parametersWithSuggestions.All(p => !p.Suggestions.Any()),
+                Parameters = parametersWithSuggestions
             };
+        }
+
+        public string GetExpression(string @namespace, ConnectorParameters parameters)
+        {
+            if (!parameters.IsCompleted)
+            {
+                return null;
+            }
+
+            string expr = $@"{@namespace}.{Name}({string.Join(", ", parameters.Parameters.Take(Math.Min(parameters.Parameters.Length, ArityMax - 1)).Select(p => p.Value.ToExpression()))}";
+
+            if (parameters.Parameters.Length > ArityMax - 1)
+            {
+                ConnectorParameterWithSuggestions lastParam = parameters.Parameters.Last();
+                expr += $@", {{ {string.Join(", ", lastParam.ParameterNames.Zip(lastParam.Values, (name, fv) => $"{name}: {fv.ToExpression()}"))} }}";
```

Comment:
>fv.ToExpression() [](http://example.com/codeflow?start=123&length=17)

we should use StringBuilder here, and then we can pass it into fv.ToExpression as well.